report_backend/time_spent_report.py

```diff
@@ -0,0 +1,138 @@
+            for course in time_spent_data.keys():
+                file_name = '{}'.format(course)
+
+                if not time_spent_data.get(course):
+                    continue
+
+                analytics_data = time_spent_data.get(course).get('analytics_data', [])
+                course_structure_data = time_spent_data.get(course).get('course_structure', [])
+
+                subsection_data = count_analytics_subsections(analytics_data, course_structure_data)
+                report_csv_headers = subsection_data[0].keys()
+
+                self.create_csv_file(file_name, subsection_data, report_csv_headers)
+
+
+    def create_csv_file(self, file_name, body_dict, headers):
+        """
+        Creates the csv file with the passed arguments, and then save it locally.
+
+        Args:
+            file_name: File string name.
+            body_dict: Dict with the data to write the csv file.
+            headers: List with the csv column names.
+        """
+        path_file = '{parent_folder}/result/{file_name}.csv'.format(
+            parent_folder=os.path.join(os.path.dirname(__file__), os.pardir),
+            file_name=file_name
+        )
+
+        with open(path_file, mode='w', encoding='utf-8') as csv_file:
+            column_headers = headers
+            writer = csv.DictWriter(csv_file, fieldnames=column_headers)
+
+            writer.writeheader()
+            for row in body_dict:
+                writer.writerow(row)
+
+        self.upload_file_to_storage(file_name, path_file)
+
+
+    def upload_file_to_storage(self, course, path_file):
+        """
+        Uploads the csv report, to S3 storage.
+
+        Args:
+            course: Course string id.
+            path_file: CSV report local file path.
+        """
+        amazon_storage = boto3.resource('s3')
+        reports_bucket = amazon_storage.Bucket('proversity-custom-reports')
+        now = datetime.now()
+
+        reports_bucket.upload_file(
+            path_file,
+            'cabinet/{course}/time_spent_report/{date}.csv'.format(
+                course=course,
+                date=now
+            )
+        )
+
+
+def count_analytics_subsections(analytics_data, course_structure_data):
+    """
+    Returns a dict containing the essential data to generate the time spent csv report.
+
+    Args:
+        analytics_data: List with the Google Analytics data.
+        course_structure_data: List with the course structure by subsection data.
+    Returns:
+        List containing the data to generate the report.
+        [{
+            section: Course section string name.
+            subsection: Course subsection string name.
+            page_views: Total of subsection views.
+            time_on_page: Total of the average time on page.
+        }]
+    """
+    subsection_data = []
+
+    for course_block in course_structure_data:
+        sequential_id = course_block.get('sequential_id', '')
+        subsections_occurrences = [
```

Comment:
This could be changed to a for loop and include  the lines 126 and 127 after an if statement in order to use the first iteration. is there a performance advantage with this code ??
```
for item in analytics_data:
    if sequential_id in item.get('page_path'):
        total_page_views += int(item.get('page_views', 0))
        total_time_on_page += float(item.get('avg_time_on_page', 0))
```

Reply:
Yes, since analytics_data has all information, and subsections_occurrences just have the ocurrences information, and the comprenhension makes the loop. @andrey-canon